### ISSUE TITLE
[Q] how to log user-defined metric

### ISSUE THREAD

**Issue by LAB-703** [wandb/wandb#5982]
Hi, I am happy to use wandb. Could you give me a tip on how to log custom metrics?

I've already tested this example and it works fine. https://docs.wandb.ai/guides/track/log/log-summary

But I can't log a custom metric in my code. Here's the code.
I don't know where "wandb.log()" is added.

`
    accuracy_score = evaluate.load('accuracy')
    precision_score = evaluate.load('precision')
    recall_score = evaluate.load('recall')
    f1_score = evaluate.load('f1')

    def compute_metrics_fn(eval_pred):    
        preds, labels = eval_pred
        preds = np.argmax(preds, axis=1)
        accuracy = accuracy_score(references=labels, predictions=preds)
        precision = precision_score(references=labels, predictions=preds, average = 'macro')
        recall = recall_score(references=labels, predictions=preds, average = 'macro')
        f1 = f1_score(references=labels, predictions=preds, average = 'macro')
        
        result = {"accuracy": accuracy, "precision": precision, "recall": recall, "f1": f1} 

        return result

    trainer = transformers.Trainer(
            model=model,
            train_dataset=train_data,
            eval_dataset=val_data,
            callbacks = [EarlyStoppingCallback(early_stopping_patience=5)],
            args=transformers.TrainingArguments(
                per_device_train_batch_size=micro_batch_size,
                gradient_accumulation_steps=gradient_accumulation_steps,
                warmup_steps=0,
                num_train_epochs=num_epochs,
                learning_rate=learning_rate,
                fp16=True,
                logging_steps=10,
                optim="adamw_torch",
                evaluation_strategy="steps", #  if val_set_size > 0 else "no",
                save_strategy="steps",
                # evaluation_strategy="epoch", # if val_set_size > 0 else "no",
                # save_strategy="epoch",
                eval_steps=50,  # if val_set_size > 0 else None,
                save_steps=50,
                output_dir=output_dir,
                save_total_limit=3,
                load_best_model_at_end=True, # if val_set_size > 0 else False,
                # metric_for_best_model='eval_f1',
                ddp_find_unused_parameters=False if ddp else None,
                group_by_length=group_by_length,
                report_to=None, # "wandb" if use_wandb else None,
                run_name=None, #wandb_run_name if use_wandb else None,
            ),
            compute_metrics=compute_metrics_fn,  ## added                       
            data_collator=transformers.DataCollatorForSeq2Seq(
                tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
            ),
        )
`

**Comment by rsanandres-wandb** [comment]
Hello @LAB-703!

Looking into this right now, I will get back to you once I find something. It looks like you will have to subclass the HuggingFace Trainer in order to add any new metrics.

**Comment by LAB-703** [comment]
Thank you. Waiting for your reply, I'll do some digging myself.

**Comment by rsanandres-wandb** [comment]
Hello! After looking around, this does not seem to be possible. However, what you can do is [add the metrics after the run is complete.](https://docs.wandb.ai/guides/technical-faq/metrics-and-performance#how-can-i-log-additional-metrics-after-a-run-completes) Since the metrics you are concerned about are accuracy, precision, recall, and f1, I believe that adding these metrics retroactively will be beneficial for you since the current workaround would be to edit the source code on your own repo to fit your needs.

**Comment by LAB-703** [comment]
Thanks for reaching out to me.
But, I got the result of new metrics from similar code.

`def compute_metrics_fn(p):
    pred, labels = p
    pred = np.argmax(pred, axis=1)

    accuracy = accuracy_score(y_true=labels, y_pred=pred)
    recall = recall_score(y_true=labels, y_pred=pred, average = 'micro')
    precision = precision_score(y_true=labels, y_pred=pred, average = 'micro')
    f1 = f1_score(y_true=labels, y_pred=pred, average = 'micro')

    return {"accuracy": accuracy, "precision": precision, "recall": recall, "f1": f1}
`
`
# Define Trainer
args = TrainingArguments(
    output_dir=f"./model/{model_name}/{datetime.now().strftime('%m-%d_%H:%M:%S')}",
    save_strategy ="steps",
    evaluation_strategy="steps",
    eval_steps=50,
    save_steps=50,
    warmup_steps=50,   
    weight_decay=0.01, 
    logging_dir='./logs',            
    logging_steps=50,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=10,		
    save_total_limit=3,
    seed=0,
    load_best_model_at_end=True,
    report_to = 'wandb'
)
trainer = Trainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics_fn,
    callbacks=[EarlyStoppingCallback(early_stopping_patience=5)],
)

# Train pre-trained model
trainer.train()`

<img width="1050" alt="image" src="https://github.com/wandb/wandb/assets/92988910/6ce5e13c-f45f-4869-91a2-1fefd576dba1">

It works well. I don't know the difference but, I'm experimenting something.
 I'll get back to you if I find something.

**Comment by sydholl** [comment]
Sydney Hollingsworth commented: 
Reported by user: LAB-703
 GitHub URL: https://github.com/wandb/wandb/issues/5982

**Comment by sydholl** [comment]
Sydney Hollingsworth commented: 
Auto-assigned

**Comment by sydholl** [comment]
Sydney Hollingsworth commented: 
This ticket has gone stale, please follow up or close.

**Comment by sydholl** [comment]
Sydney Hollingsworth commented: 
This ticket has gone stale, please follow up or close.

**Comment by sydholl** [comment]
Raphael San Andres commented: 
User found solution, closing.
